src/test/protocol/KnownFormatToGRPC_test.cpp

```diff
@@ -0,0 +1,924 @@
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/safe_cast.h>
+#include <ripple/beast/unit_test.h>
+#include <ripple/protocol/InnerObjectFormats.h>
+#include <ripple/protocol/LedgerFormats.h>
+#include <ripple/protocol/TxFormats.h>
+
+#include "org/xrpl/rpc/v1/ledger_objects.pb.h"
+#include "org/xrpl/rpc/v1/transaction.pb.h"
+
+#include <cctype>
+#include <map>
+#include <string>
+#include <type_traits>
+
+namespace ripple {
+
+// This test suite uses the google::protobuf::Descriptor class to do runtime
+// reflection on our gRPC stuff.  At the time of this writing documentation
+// for Descriptor could be found here:
+//
+// https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.descriptor#Descriptor
+
+class KnownFormatToGRPC_test : public beast::unit_test::suite
+{
+private:
+    static constexpr auto fieldTYPE_UINT32 =
+        google::protobuf::FieldDescriptor::Type::TYPE_UINT32;
+
+    static constexpr auto fieldTYPE_UINT64 =
+        google::protobuf::FieldDescriptor::Type::TYPE_UINT64;
+
+    static constexpr auto fieldTYPE_BYTES =
+        google::protobuf::FieldDescriptor::Type::TYPE_BYTES;
+
+    static constexpr auto fieldTYPE_STRING =
+        google::protobuf::FieldDescriptor::Type::TYPE_STRING;
+
+    static constexpr auto fieldTYPE_MESSAGE =
+        google::protobuf::FieldDescriptor::Type::TYPE_MESSAGE;
+
+    // Format names are CamelCase and FieldDescriptor names are snake_case.
+    // Convert from CamelCase to snake_case.  Do not be fooled by consecutive
+    // capital letters like in NegativeUNL.
+    static std::string
+    formatNameToEntryTypeName(std::string fmtName)
+    {
+        std::string entryName;
+        entryName.reserve(fmtName.size());
+        bool prevUpper = false;
+        for (std::size_t i = 0; i < fmtName.size(); i++)
+        {
+            char const ch = fmtName[i];
+            bool const upper = std::isupper(ch);
+            if (i > 0 && !prevUpper && upper)
+                entryName.push_back('_');
+
+            prevUpper = upper;
+            entryName.push_back(std::tolower(ch));
+        }
+        return entryName;
+    };
+
+    // Create a map of (most) all the SFields in an SOTemplate.  This map
+    // can be used to correlate a gRPC Descriptor to its corresponding SField.
+    template <typename KeyType = int>
```

Comment:
The template parameter is going to be inferred from the function argument. I don't think the default parameter will be used. I'd remove it (unless there's a rational for it that I'm missing).